base/shell/explorer/desktop.cpp

```diff
@@ -22,107 +22,160 @@
 
 class CDesktopThread
 {
-    HANDLE m_hEvent;
+private:
+    
     CComPtr<ITrayWindow> m_Tray;
+    
+    HANDLE m_hEvent;
+    HANDLE m_hThread;
 
-    DWORD DesktopThreadProc()
-    {
-        CComPtr<IShellDesktopTray> pSdt;
-        HANDLE hDesktop;
-        HRESULT hRet;
-
-        OleInitialize(NULL);
-
-        hRet = m_Tray->QueryInterface(IID_PPV_ARG(IShellDesktopTray, &pSdt));
-        if (!SUCCEEDED(hRet))
-            return 1;
+    DWORD DesktopThreadProc();
+    static DWORD WINAPI s_DesktopThreadProc(LPVOID lpParameter);
 
-        hDesktop = _SHCreateDesktop(pSdt);
-        if (hDesktop == NULL)
-            return 1;
+public:
+    
+    CDesktopThread();
+    virtual ~CDesktopThread();
 
-        if (!SetEvent(m_hEvent))
-        {
-            /* Failed to notify that we initialized successfully, kill ourselves
-            to make the main thread wake up! */
-            return 1;
-        }
+    HRESULT Initialize(ITrayWindow* pTray);
+    void Destroy();
+};
 
-        _SHDesktopMessageLoop(hDesktop);
+/*******************************************************************/
 
-        /* FIXME: Properly rundown the main thread! */
-        ExitProcess(0);
+CDesktopThread::CDesktopThread()
+{
+    m_Tray = NULL;
+    m_hEvent = NULL;
+    m_hThread = NULL;
+}
 
-        return 0;
+CDesktopThread::~CDesktopThread()
+{
+    if (m_hThread)
+    {
+        CloseHandle(m_hThread);
     }
 
-    static DWORD CALLBACK s_DesktopThreadProc(IN OUT LPVOID lpParameter)
+    if (m_hEvent)
     {
-        return reinterpret_cast<CDesktopThread*>(lpParameter)->DesktopThreadProc();
+        CloseHandle(m_hEvent);
     }
+}
 
-public:
-    CDesktopThread() :
-        m_hEvent(NULL),
-        m_Tray(NULL)
+HRESULT CDesktopThread::Initialize(ITrayWindow* pTray)
+{
+    HANDLE Handles[2];
+
+    m_Tray = pTray;
+
+    m_hEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
+
+    if (!m_hEvent)
     {
+        return E_FAIL;
     }
 
-    HRESULT Initialize(IN OUT ITrayWindow *pTray)
-    {
-        HANDLE hThread;
-        HANDLE Handles[2];
+    m_hThread = CreateThread(NULL, 0, s_DesktopThreadProc, (LPVOID)this, 0, NULL);
+
+    if (!m_hThread)
+    {   
+        return E_FAIL;
```

Comment:
The function creates a handle, but then something else fails: it needs to do its own cleanup.
```suggestion
        CloseHandle(m_hEvent);
        m_hEvent = NULL;
        return E_FAIL;
```
Because otherwise, what if somebody who uses the CDesktopThread() does:
```
ret = someThread->Initialize(tray1);
if (ret != S_OK)
{
    ret = someThread->Initialize(tray2);
}
```
?
You would get leakage of the event handle.

Reply:
That's invalid usage for every `Initialize` function I've ever seen. It'd be easier to just assert `m_hEvent == nullptr` at the beginning of `Initialize` or something.